test/libsolidity/semanticTests/using/library_functions_inside_contract.sol

```diff
@@ -0,0 +1,30 @@
+library L {
+    function publicFunction(uint a) public pure returns (uint) { return a * 1; }
+    function externalFunction(uint b) external pure returns (uint) { return b * 2; }
+    function internalFunction(uint c) internal pure returns (uint) { return c * 3; }
+}
+
+contract C {
+    using {L.publicFunction} for uint;
+    using {L.externalFunction} for uint;
+    using {L.internalFunction} for uint;
+
+    function f() public pure returns (uint) {
+        uint x = 1;
+        return x.publicFunction();
+    }
+
+    function g() public pure returns (uint) {
+        uint x = 1;
+        return x.externalFunction();
+    }
+
+    function h() public pure returns (uint) {
+        uint x = 1;
+        return x.internalFunction();
+    }
+}
+// ----
+// f() -> FAILURE
+// g() -> FAILURE
```

Comment:
I am trying to understand why these resulted in failure.

Reply:
Try adding `// library: L` before the calls!

Reply:
The explanation is the following:
These are external library calls, which are a special kind of "contract-to-contract" calls, i.e. calls to a different address on chain (in particular a ``delegatecall``, i.e. a call that executes code at some other address, while still working with the storage of the calling contract).

In our generated bytecode, the addresses called in the generated ``delegatecall``s for these external libraries are placeholders (resp. linker symbols). Before actually deploying such a contract, it usually has to be "linked" first, which means filling in the addresses at which the libraries have been previously deployed on chain.

In our test suite, deploying a library is done with the special ``// library: <name>`` comment and linker symbols for any contracts after that are automatically resolved.

Reply:
Without the ``// library: L``, the linker references remain unresolved, and the library calls end attempting to call into nowhere, causing them to revert in panic -> FAILURE.